core/apple/src/SinksApple.kt

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2017-2023 JetBrains s.r.o. and respective authors and developers.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the LICENCE file.
+ */
+
+package kotlinx.io
+
+import kotlinx.cinterop.*
+import platform.Foundation.*
+import platform.darwin.NSInteger
+import platform.darwin.NSUInteger
+import platform.posix.memcpy
+import platform.posix.uint8_tVar
+import kotlin.concurrent.Volatile
+
+/**
+ * Returns an output stream that writes to this sink. Closing the stream will also close this sink.
+ *
+ * @sample kotlinx.io.samples.KotlinxIoSamplesApple.asStream
+ */
+public fun Sink.asNSOutputStream(): NSOutputStream = SinkNSOutputStream(this)
+
+@OptIn(UnsafeNumber::class)
+private class SinkNSOutputStream(
+    private val sink: Sink,
+) : NSOutputStream(toMemory = Unit) {
+
+    private val isClosed: () -> Boolean = when (sink) {
+        is RealSink -> sink::closed
+        is Buffer -> {
+            { false }
+        }
+    }
+
+    @OptIn(ExperimentalStdlibApi::class)
+    @Volatile
+    private var status = NSStreamStatusNotOpen
```

Comment:
Why is the status volatile?
It seems like NSStream doesn't provide any thread-safety guarantees.

Reply:
Yes, despite not providing thread safety guarantees for operating on the streams, I just didn't see much use for the status being updated to `NSStreamStatusReading` and `NSStreamStatusWriting` if this wasn't at least visible to another thread from a read-only perspective. The thread doing the reading and writing would never see the status change, since it's in the middle of the read and write functions.

I suppose a caller could use key-value coding to observe the `streamStatus` property and the observer method would be called on the same thread that updated the status in the middle of the read or write. This might actually be the only really valid use case for those statuses, since it's likely not feasible to poll the status from another thread.

Are there reasons to not make the property volatile? If so, should I remove the `@Volatile` annotation then?

Reply:
That makes sense. I'll remove the `@Volatile` annotations.